src/Geocoder/Provider/AbstractHttpProvider.php

```diff
@@ -18,27 +21,28 @@
 class AbstractHttpProvider extends AbstractProvider
 {
     /**
-     * @var HttpAdapterInterface
+     * @var HttpClient
      */
-    private $adapter;
+    private $client;
 
     /**
-     * @param HttpAdapterInterface $adapter An HTTP adapter
+     * @param HttpClient $client An HTTP adapter
      */
-    public function __construct(HttpAdapterInterface $adapter)
+    public function __construct(HttpClient $client = null, MessageFactory $factory = null)
     {
         parent::__construct();
 
-        $this->adapter = $adapter;
+        $this->client = $client ?: HttpClientDiscovery::find();
+        $this->factory = $factory ?: MessageFactoryDiscovery::find();
```

Comment:
I'm not sure about the discovery service at all, it brings many problems and puli as new dependency. Maybe we should wait for a stable release of the discovery service.